File: preIn.py
#Time Complexity: O(n^2)
#Space Complexity: O(n^2)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        
        if len(preorder) == 0:
            return None
        rootVal = preorder[0]
        root = TreeNode(rootVal)
        rootIdx = -1
        
        for i in range(len(inorder)):
            if inorder[i] == rootVal:
                rootIdx = i
                break
        
        inLeft = inorder[:rootIdx]
        inRight = inorder[rootIdx+1:]
        preLeft = preorder[1:len(inLeft)+1]
        preRight = preorder[len(inLeft)+1:]
        
        root.left = self.buildTree(preLeft, inLeft)
        root.right = self.buildTree(preRight, inRight)
        return root
        
        
        
        
        

File: Pre-In.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    int rdx = 0; // running on the preorder
         map<int, int> m;
   
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {

     
        int n = inorder.size();

        TreeNode* ans = pre(preorder,inorder,0,n-1,n);
        return ans;

    }

    int findpos(vector<int>& inorder,int rootVal,int n){
        for(int i = 0;i<n ; i++){
            if(rootVal == inorder[i]){
                return i;
            }
   
        }
        return -1;
    }

    TreeNode* pre(vector<int>& preorder, vector<int>& inorder, int s,int e,int size){
            if(s>e || rdx >= size){
                return NULL;
            }

            int rootVal = preorder[rdx];
            rdx++;
            TreeNode* root = new TreeNode(rootVal);
            int rootIdx = findpos(inorder,rootVal,size);
            root->left = pre(preorder,inorder,s,rootIdx-1,size);
            root->right = pre(preorder,inorder,rootIdx+1,e,size);

      return root;
    }
};